File: algorithms/lua/kmp.lua
local function build_lps(pat, m)
    local lps = {}
    for i = 1, m do lps[i] = 0 end
    local len, i = 0, 2
    while i <= m do
        if pat:sub(i,i) == pat:sub(len+1,len+1) then
            len = len + 1
            lps[i] = len
            i = i + 1
        elseif len > 0 then
            len = lps[len]
        else
            lps[i] = 0
            i = i + 1
        end
    end
    return lps
end

local function kmp(text, pat)
    local n, m = #text, #pat
    if m == 0 then return {} end
    local lps     = build_lps(pat, m)
    local matches = {}
    local i, j   = 1, 1
    while i <= n do
        if text:sub(i,i) == pat:sub(j,j) then
            i = i + 1; j = j + 1
        end
        if j > m then
            table.insert(matches, i - j - 1)  -- 0-based
            j = lps[j-1] + 1
        elseif i <= n and text:sub(i,i) ~= pat:sub(j,j) then
            if j > 1 then j = lps[j-1] + 1 else i = i + 1 end
        end
    end
    return matches
end

local function main()
    local text = io.read("*l")
    local pat  = io.read("*l")

    local start   = os.clock()
    local matches = kmp(text, pat)
    local ms      = (os.clock() - start) * 1000

    print(#matches)
    if #matches == 0 then
        print("")
    else
        print(table.concat(matches, " "))
    end
    io.stderr:write(string.format("%.3f\n", ms))
end

main()

File: algorithms/haskell/Kmp.hs
import Data.Array
import Data.Time.Clock
import System.IO
import Control.Exception (evaluate)

buildLPS :: String -> Array Int Int
buildLPS pat =
    let m   = length pat
        p   = listArray (0, m-1) pat
        lps = array (0, m-1) $ go 0 1 []
    in lps
  where
    go len i acc
        | i >= length pat = acc
        | pat !! i == pat !! len =
            let len' = len + 1
            in go len' (i+1) ((i, len') : acc)
        | len > 0 = go (buildLPS pat ! (len-1)) i acc
        | otherwise = go 0 (i+1) ((i, 0) : acc)

kmp :: String -> String -> [Int]
kmp text pat
    | null pat = []
    | otherwise =
        let n   = length text
            m   = length pat
            t   = listArray (0, n-1) text
            p   = listArray (0, m-1) pat
            lps = buildLPS pat
        in go t p lps n m 0 0 []
  where
    go t p lps n m i j acc
        | i >= n = reverse acc
        | t ! i == p ! j =
            let i' = i+1; j' = j+1
            in if j' == m
               then go t p lps n m i' (lps ! (j'-1)) ((i'-j') : acc)
               else go t p lps n m i' j' acc
        | j > 0  = go t p lps n m i (lps ! (j-1)) acc
        | otherwise = go t p lps n m (i+1) 0 acc

main :: IO ()
main = do
    hSetBuffering stdout LineBuffering
    contents <- getContents
    let ls  = lines contents
        text = head ls
        pat  = ls !! 1

    start   <- getCurrentTime
    matches <- evaluate (kmp text pat)
    end     <- getCurrentTime

    let ms = realToFrac (diffUTCTime end start) * 1000 :: Double
    print (length matches)
    if null matches
        then putStrLn ""
        else putStrLn $ unwords (map show matches)
    hPutStrLn stderr $ show ms